src/overlays/actors/ovl_En_Slime/z_en_slime.c

```diff
@@ -88,103 +128,1169 @@ static DamageTable D_80A31B1C = {
+            EnSlime_SetIceBlockedState(this);
+        } else {
+            this->actor.colorFilterTimer = 0;
+            this->collider.base.acFlags |= AC_ON;
+            this->iceBlockTimer = 255;
+            this->actor.flags |= ACTOR_FLAG_1;
+            this->actor.gravity = -2.0f;
+            EnSlime_ReturnToIdle(this);
+        }
+    }
+}
+
+/**
+ * Enter state as a chuchu frozen in an ice block.
+ * This function is called after ice block spawn is successful.
+ * Action function set to UpdateIceBlockedState
+ */
+void EnSlime_SetIceBlockedState(EnSlime* this) {
+    this->actor.flags |= ACTOR_FLAG_10;
+    this->actionFunc = EnSlime_UpdateIceBlockedState;
+}
+
+/**
+ * Decrement the ice block timer, or if it has hit zero, check to
+ * see if the ice block child actor link has been broken or if it has
+ * shrunk sufficiently.
+ * If timer is up and ice block is sufficiently "gone", init unfreeze.
+ */
+void EnSlime_UpdateIceBlockedState(EnSlime* this, PlayState* play) {
+    if (this->iceBlockTimer == 0) {
+        // Proceed to next action if ice block link or its update link are broken or the ice block has shrunk
+        // sufficiently.
+        if ((this->actor.child == NULL) || (this->actor.child->update == NULL) ||
+            !(this->actor.child->scale.y >= 0.1f)) {
+            EnSlime_InitIceBlockUnfreeze(this);
+        }
+    } else {
+        this->actor.colorFilterTimer = 10;
+        if ((this->iceBlockTimer - 5) < 0) {
+            this->iceBlockTimer = 0;
+        } else {
+            this->iceBlockTimer -= 5;
+        }
+    }
+}
+
+/**
+ * Prepare the actor to behave as if stunned. Set speed to zero,
+ * and set action function to UpdateStunnedState.
+ */
+void EnSlime_EnterStunnedState(EnSlime* this) {
+    this->actor.speed = 0.0f;
+    func_800BE504(&this->actor, &this->collider);
+    if (this->actor.velocity.y > 0.0f) {
+        this->actor.velocity.y = 0.0f;
+    }
+    this->actionFunc = EnSlime_UpdateStunnedState;
+}
 
-extern ColliderCylinderInit D_80A31AF0;
-extern DamageTable D_80A31B1C;
-extern CollisionCheckInfoInit D_80A31B3C;
-extern InitChainEntry D_80A31B54[];
+/**
+ * Check the actor's timer to see whether stun has worn off.
+ * If so, either react to damage and initialize death, or start
+ * idly wobbling in a new direction.
+ */
+void EnSlime_UpdateStunnedState(EnSlime* this, PlayState* play) {
+    this->timer--;
+    if (this->timer == 0) {
+        EnSlime_UpdateIceEffects(this, play);
+        if (this->actor.colChkInfo.health == 0) {
+            EnSlime_ReactToDamage(this, play, false);
+        } else {
+            this->actor.world.rot.y = this->actor.shape.rot.y;
+            EnSlime_ChangeDirectionIdle(this);
+        }
+    }
+}
 
-extern UNK_TYPE D_060004C0;
-extern UNK_TYPE D_06000828;
+/**
+ * Setup ice block unfreeze
+ * Set action function to UpdateIceBlockUnfreeze
+ */
+void EnSlime_InitIceBlockUnfreeze(EnSlime* this) {
```

Comment:
Another setup function; I would do something like `EnSlime_SetupIceBlockThaw` and `EnSlide_IceBlockThaw`, but you can think of something too. I just know that we prefer "thaw" to "unfreeze".